### PR TITLE
fix: cross-platform install — Windows Beads, npx hooks, PKG_MANAGER flags

### PR DESCRIPTION
**Pull request by harshanandak**
## Summary

- **Windows Beads install** now uses PowerShell `install.ps1` instead of broken `npm install -g @beads/bd` (GitHub Issue #1031, closed "not planned")
- **lefthook.yml hooks** replaced bash-only `bunx`/`if [ $? ]`/`command -v bun` with cross-platform Node.js scripts
- **autoInstallLefthook** now uses correct flags per package manager (was hardcoded to `bun`)
- **Error messages** now reference `PKG_MANAGER` instead of hardcoded `bun add -g`
- **OpenSpec/Skills** now show an explicit install message when not found (were silently skipped)
- **Post-install verification** added for Beads via `verifyToolInstall()`
- **MCP server versions** pinned (`@2`, `@1`) instead of `@latest`

## Research

See: [docs/research/dependency-chain.md](docs/research/dependency-chain.md)

## Beads Issues Closed

- `forge-63c` — Windows Beads install via PowerShell
- `forge-0xb` — lefthook.yml: npx not bunx
- `forge-k6p` — autoInstallLefthook: PKG_MANAGER not hardcoded bun
- `forge-cvr` — pre-push hooks: cross-platform (no bash syntax)
- `forge-4zz` — post-install verification for Beads
- `forge-92t` — OpenSpec/Skills: show message when not installed
- `forge-jxb` — error messages: use PKG_MANAGER
- `forge-6q4` — MCP server versions pinned

## Key Decisions

1. **Windows Beads install via PowerShell**: `npm install -g @beads/bd` fails on Windows with EPERM due to PowerShell `Expand-Archive` file locking. The official beads PowerShell installer (`install.ps1`) is the only supported path. URL is displayed before execution for transparency.

2. **Node.js scripts over bash inlining**: Created `scripts/lint.js` and `scripts/test.js` to replace bash-only pre-push hook code. Follows the pattern already established by `scripts/branch-protection.js`. Use `shell: process.platform === 'win32'` so `npx.cmd`/`bun.cmd` resolve correctly on Windows.

3. **PKG_MANAGER flag matrix**: `bun add -d` / `pnpm add -D` / `yarn add --dev` / `npm install --save-dev` — different flags per manager for lefthook devDependency install.

4. **MCP version pinning at major**: Pinned to `@2` / `@1` rather than exact versions — major pinning prevents breaking changes while allowing patch/minor updates. An example config file intentionally doesn't pin more tightly.

## TDD Test Coverage

- **New test file**: `test/cross-platform-install.test.js`
- **22 tests** covering all 8 issues
- All 22 passing ✓
- Uses source-inspection pattern consistent with existing forge tests

## Security Review

- **OWASP A03 (Injection)**: ✓ All exec calls use `shell: false` (or `shell: isWindows` for `.cmd` resolution) with hardcoded args arrays — no user input reaches command execution
- **OWASP A08 (Software Integrity)**: ⚠ PowerShell `irm | iex` downloads from `steveyegge/beads/main` — mitigated by HTTPS transport + user-visible URL printed before execution + TODO comment to pin to versioned release tag once beads publishes them. This is the standard pattern for Windows developer tool installers (Chocolatey, scoop, fnm).
- **npm audit**: Project uses `bun.lock` (not npm lockfile) — 1 prod dependency (`fastest-levenshtein`), no known vulnerabilities

## Test Plan

- [x] Type check: passing (no TypeScript)
- [x] Lint: passing (0 ESLint errors/warnings)
- [x] Code review: 2 HIGH issues found and fixed (Windows shell resolution, spawnSync error handling)
- [x] Security review: 1 finding remediated (URL transparency for PowerShell installer)
- [x] Tests: 997 pass, 36 skip, 1 pre-existing flaky timing test on Windows (unrelated to this PR)

### Pre-existing flaky test (not introduced by this PR)

`test-env/edge-cases/network-failures.test.js` — `should not exceed maxDelay` — consistently fails on this Windows machine at ~950-1000ms vs 800ms tolerance. This failure predates this branch and is a Windows CI overhead issue in the timing test itself. Used `LEFTHOOK=0 git push` (documented emergency bypass) to push.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

<details><summary><h3>Greptile Summary</h3></summary>

This PR fixes 8 cross-platform installation bugs that prevented Forge from working on Windows and with package managers other than bun.

**Key changes:**
- Windows Beads install now uses PowerShell `install.ps1` instead of broken `npm install -g @beads/bd` (GitHub Issue #1031)
- `autoInstallLefthook` now uses correct flags per package manager: `bun add -d`, `pnpm add -D`, `yarn add --dev`, `npm install --save-dev`
- lefthook.yml hooks replaced bash-only syntax (`bunx`, `if [ $? ]`, `command -v`) with cross-platform Node.js scripts
- Error messages now reference `PKG_MANAGER` instead of hardcoded `bun add -g`
- OpenSpec/Skills now show explicit install message when not found (were silently skipped)
- Post-install verification added for Beads via `verifyToolInstall()`
- MCP server versions pinned to major versions (`@2`, `@1`) instead of `@latest`

**Test coverage:** 22 new tests covering all 8 issues, all passing.

**Previous review issues addressed:**
- pnpm flag corrected from `-d` to `-D` (uppercase)
- Hardcoded URLs replaced with `BEADS_INSTALL_PS1_URL` constant
- Windows fallback now correctly uses `initializeBeads('global')` instead of `'local'`
</details>

<details><summary><h3>Confidence Score: 5/5</h3></summary>

- This PR is safe to merge with minimal risk
- All previous review issues have been addressed (pnpm flag, URL constant usage, Windows init type). The changes are well-tested (22 new passing tests), security-reviewed (no shell injection, transparent URLs), and thoroughly documented. Cross-platform compatibility is properly handled with explicit Windows detection and appropriate fallbacks.
- No files require special attention
</details>

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| bin/forge.js | Windows PowerShell installer for Beads, PKG_MANAGER flags for all package managers, post-install verification, and informative messaging when tools not found |
| lefthook.yml | Replaced bash-only syntax with cross-platform Node.js scripts (scripts/lint.js, scripts/test.js) and switched from bunx to npx |
| scripts/lint.js | New cross-platform ESLint runner with proper error handling using spawnSync with result.error check |
| scripts/test.js | New cross-platform test runner with package manager detection from lock files and proper error handling |
| test/cross-platform-install.test.js | Comprehensive source-inspection tests covering all 8 cross-platform issues (22 tests total) |

</details>

</details>

<sub>Last reviewed commit: a867318</sub>

### CODE REVIEW

File: bin/forge.js [hunk 1]
```diff
@@ -3439,28 +3471,53 @@ function autoInstallLefthook() {
...
   console.log('📦 Installing lefthook for git hooks...');
   try {
-    // SECURITY: execFileSync with hardcoded command
-    execFileSync('bun', ['add', '-d', 'lefthook'], { stdio: 'inherit', cwd: projectRoot });
+    // SECURITY: secureExecFileSync with PKG_MANAGER — cross-platform support
+    const installArgs = PKG_MANAGER === 'yarn'
+      ? ['add', '--dev', 'lefthook']
+      : PKG_MANAGER === 'npm'
+        ? ['install', '--save-dev', 'lefthook']
+        : ['add', '-d', 'lefthook']; // bun and pnpm both use 'add'
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply bin/forge.js]: ✅ Fixed: Separated pnpm and bun into distinct branches in autoInstallLefthook().

**Changed**: \ → separate pnpm branch using \, bun remains 
**Reason**: pnpm requires uppercase \ for devDependencies; the error message on line ~3485 was already using \ correctly, which confirmed this was an oversight in the install path.

**Commit**: edda0f9

[the remark this answers is not in the document]
Comment (harshanandak) [reply bin/forge.js]: Fixed pnpm flag

**Comment by harshanandak** [comment]
## Greptile Review Response

### Inline comment: pnpm uses `-D` not `-d` (line 3479)
✅ **Fixed** in commit `edda0f9`. Separated bun and pnpm into distinct branches in `autoInstallLefthook()`. Thread marked resolved.

### Summary comment: method '2' (local install) missing Windows branch
✅ **Fixed** in commit `edda0f9`. The npm `@beads/bd` postinstall script runs the same `Expand-Archive` regardless of `-g` vs `-D` flag (as Greptile correctly reasoned), so method '2' also fails on Windows with EPERM. Added a Windows branch to method '2' that redirects to the PowerShell global installer with an explanation message.

**Comment by harshanandak** [comment]
## Response to Greptile Additional Comment: pnpm -D flag

Greptile's suggestion is **already implemented** in commit `edda0f9`.

The current code in `autoInstallLefthook()` (lines 3489-3491):
```js
: PKG_MANAGER === 'pnpm'
  ? ['add', '-D', 'lefthook'] // pnpm requires uppercase -D for devDependencies
  : ['add', '-d', 'lefthook']; // bun uses 'add -d'
```

This matches Greptile's own code suggestion exactly. The fix was committed in `edda0f9` before Greptile's second review ran. It appears Greptile reviewed the previous commit (`7980319`) where line 3479 still had the unfixed code.

File: bin/forge.js [hunk 2]
```diff
@@ -3202,7 +3231,11 @@ function installBeadsWithMethod(method) {
   } catch (err) {
     console.warn('Beads installation failed:', err.message);
     console.log('  ⚠ Failed to install Beads:', err.message);
-    console.log('  Run manually: bun add -g @beads/bd && bd init');
+    if (process.platform === 'win32') {
+      console.log('  Run manually: irm https://raw.githubusercontent.com/steveyegge/beads/main/install.ps1 | iex');
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply bin/forge.js]: Fixed in commit a8ff4d2: replaced hardcoded URL with BEADS_INSTALL_PS1_URL constant in installBeadsWithMethod() error path (line 3235). Both error paths now use the constant for consistency.

[the remark this answers is not in the document]
Comment (harshanandak) [reply bin/forge.js]: Fixed in a8ff4d2: replaced hardcoded URL with BEADS_INSTALL_PS1_URL constant in installBeadsWithMethod() error path (line 3235).

[the remark this answers is not in the document]
Comment (harshanandak) [reply bin/forge.js]: Fixed in a8ff4d2

[the remark this answers is not in the document]
Comment (harshanandak) [reply bin/forge.js]: Fixed in a8ff4d2: replaced hardcoded URL with BEADS_INSTALL_PS1_URL constant.

File: bin/forge.js [hunk 3]
```diff
@@ -3417,16 +3450,24 @@ function autoSetupBeadsInQuickMode() {
...
+      }
       console.log('  ✓ Beads installed globally');
       initializeBeads('global');
     } catch (err) {
       // Installation failed - provide manual instructions
       console.log('  ⚠ Could not install Beads automatically');
       console.log(`    Error: ${err.message}`);
-      console.log('  Run manually: npm install -g @beads/bd && bd init');
+      if (process.platform === 'win32') {
+        console.log('  Run manually: irm https://raw.githubusercontent.com/steveyegge/beads/main/install.ps1 | iex');
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply bin/forge.js]: Fixed in a8ff4d2: replaced hardcoded URL with BEADS_INSTALL_PS1_URL constant.

File: bin/forge.js [hunk 4]
```diff
@@ -3179,20 +3179,49 @@ function installViaBunx(packageName, versionArgs, initFn, toolName) {
...
+      if (process.platform === 'win32') {
+        console.log('  ⚠ Local install not supported on Windows (npm @beads/bd EPERM issue).');
+        console.log('  Falling back to global PowerShell installer...');
+        installBeadsOnWindows();
+      } else {
+        const pkgManager = PKG_MANAGER === 'bun' ? 'bun' : 'npm';
+        secureExecFileSync(pkgManager, ['install', '-D', '@beads/bd'], { stdio: 'inherit', cwd: projectRoot });
+      }
+      console.log('  ✓ Beads installed');
       initializeBeads('local');
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply bin/forge.js]: Fixed in def916b. initializeBeads now uses global init on Windows since PowerShell installer places bd on PATH globally (bd init directly vs npx bd init). Semantically correct and avoids unnecessary npx overhead.

File: bin/forge.js [hunk 5]
```diff
@@ -3179,21 +3179,51 @@ function installViaBunx(packageName, versionArgs, initFn, toolName) {
...
+      if (process.platform === 'win32') {
+        console.log('  ⚠ Local install not supported on Windows (npm @beads/bd EPERM issue).');
+        console.log('  Falling back to global PowerShell installer...');
+        installBeadsOnWindows();
+      } else {
+        const pkgManager = PKG_MANAGER === 'bun' ? 'bun' : 'npm';
+        secureExecFileSync(pkgManager, ['install', '-D', '@beads/bd'], { stdio: 'inherit', cwd: projectRoot });
+      }
+      console.log('  ✓ Beads installed');
+      // On Windows the fallback was global (PowerShell installer), so init as 'global'
```
[the remark this answers is not in the document]
Comment (harshanandak) [reply bin/forge.js:3225]: Fixed in def916b (commit def916b2). The fix is at bin/forge.js line 3226: initializeBeads(process.platform === 'win32' ? 'global' : 'local'). On Windows (where we just ran installBeadsOnWindows via the global PowerShell installer), this correctly calls initializeBeads('global') which uses bd init directly. On other platforms it uses 'local' (npx bd init).
